linux intro module - closes #634 (#641)
bob-crutchley/courseware: topics/linux/modules/introduction/README.md

# Linux Introduction

<!--TOC_START-->
## Contents
- [Overview](#overview)
	- [Anatomy of the OS](#anatomy-of-the-os)
	- [Linux](#linux)
	- [History of Linux](#history-of-linux)
	- [Linux & Windows](#linux--windows)
	- [Uses of Linux](#uses-of-linux)
	- [Linux Distributions](#linux-distributions)
	- [Linux Software Packages](#linux-software-packages)
- [Tutorial](#tutorial)
- [Exercises](#exercises)

<!--TOC_END-->
## Overview

> Linux is an operating system (`OS`). Think of the operating system like the controller of a computer. It is a piece of software that works with the hardware of the computer.

Reviewer: ```suggestion
Linux is an **O**perating **S**ystem (`OS`). You can think of OS as an application that takes control of the hardware.
```

### Anatomy of the OS

> The Operating System is comprised of a number of different components.

Reviewer: ```suggestion
The Operating System is comprised of a number of different components:
```

- Bootloader

> - Kernal

Reviewer: ```suggestion
- Kernel
```

- Daemon(s)
- Shell
- Graphical (X) Server
- Desktop Environment
- Applications

Since the `OS` manages all of the communication between the software and hardware. Without it, nothing can run.

**Bootloader**

This is the program that starts when the device is turned on.
It boots up the `OS`.
Typically user doesn't need to interact with the Bootloader.

If a user has more than one `OS` on their device, the user chooses on start up which Operating System they want to use and then works with that one.

> **Kernal**

Reviewer: ```suggestion
**Kernel**
```

This is the core of the system.

> The kernal manages all resources, the CPU, memory etc.

Reviewer: replace all the "karnal" to "kernel" within the file if there's more pls

It has complete control over the system.

Users can use programs that interact with the kernal.

**Daemons**
These are background processes that are ran for the `OS`.

> A process is a program being ran.

Reviewer: ```suggestion
A process is a program that is running.
```

These processes are typically started by the Operating System but can be started by the user.

The `Daemon` processes have many uses.
Including logging, managing connections and scheduling.

The first process started on a Linux system is called `system`.
It was previously known as `init`.
All other processes are derivatives of this process.

> This is started by the kernal. If it doesn't, a kernal panic will occur.

Reviewer: explain what kernel panic is as it sounds a bit ridiculous otherwise :D

**The Shell**

A user normally interacts with the system through the `shell`.
This is the process that allows users to type in commands to run.
It is completely `text based`.

The shell will be used quite extensively throughout these modules.

> There will be times where you need to work with a Linux machine and you won't have access to a `Desktop`.

Reviewer: ```suggestion
There will be times where you need to work with a Linux machine and you won't have access to a `Guided User Interface`.
```

In this case you will have a shell to work with.

**Graphical Server**

Typically refered to as the `X Window`, `X11` or `X`.

> It provides the graphical user interface for a Linux system.

Reviewer: ```suggestion
It provides the **G**raphical **U**ser **I**nterface (GUI) for a Linux system.
```

Window display, mouse movement, etc. are all handled by the `Graphical Server`.

The Graphical Server is just the baseline for a `GUI`.
Other programs will adjust the looks of the GUI.
`GNOME` is a popular example of one of these programs.
The Graphical Server simply hosts the desktop environment.

**Desktop Environment**

This is the part of the system that user will generally interact with. It is made up of programs running on top of an `OS`.

A desktop environment is normally made up of:
- Windows and toolbars
- Folders
- Desktop Icons and Widgets
- Applications

> Typically server `OS's` do not have a desktop environment.

Reviewer: ```suggestion
Typically server `OS's` do not have a desktop environment and you will only have a shell to interact with.
```

### Linux

Linux is an `open source` kernal.
Most people have used Windows on their computer, but Linux is extremely popular and is seen in a number of systems.

Linux has a massive `community` with lots of distributions being open source (free) and there is a lot of support around the `OS` and how it works.

> 

Reviewer: ```suggestion

Due to it's high efficiency majority of world super computers are running Linux.

```

### History of Linux

Linux was developed by `Linus Torvalds` in the 90's.
He developed a kernal that was to be used alongside Minix, a non-free system similar to Unix.

> It was released in 1991, and eventually combined with GNU to produce a free operating system.

Reviewer: ```suggestion
It was released in 1991, and eventually combined with GNU to produce a free operating system.
Linus is still actively contributing to the development of Linux and any change has to be approved by him, before it ends up in the next version of kernel.
```

GNU - Another free operating system as well as a collection of software. Can work alongside Linux.

### Linux & Windows

Linux and Windows have many differences.
 - Linux is free of charge and community driven.
 - The kernal is customisable
 - It's open source meaning the community is there for support
 - Application installation is centralised

### Uses of Linux

The Linux Kernal is very versatile. It can be installed on anything that has storage space and a CPU, with some modifications.

Examples of Linux in strange places:
- The Linux motorcycle - Mavizen's TTX02 / 2013 Cadillac XTS
- The Linux Coffee Pot / Smart Slow Cooker
- US Navy Sonar Systems
- The Large Hadron Collider
- The International Space Station
- Air Traffic Control Systems

### Linux Distributions

There are a variety of different versions of Linux, known as `distributions`.
All these distributions are built on the Linux kernal, but can have minor or major differences.
Some distributions are tailored for desktop use, whilst others are more commonly used for servers.

### Linux Software Packages

`Linux` normally works with packages, which are bundles of software.

> There are a number of packages on `Linux` by default, but you can install your own packages.

Reviewer: ```suggestion
There are a number of packages available on `Linux`, but you can install your own packages too.
```

> Linux uses a `package manager`. 

Reviewer: ```suggestion
Linux uses a `package manager` as a tool to do all the hard work of installing, updating and uninstalling the packages.
```

> Essentially, this tool will do all the hard work of installing, updating and uninstalling the `package`. 

Reviewer: ```suggestion
```

> All we have to do is write the command.

Reviewer: ```suggestion
All we have to do is write the command, for example installing Java JDK would be a command like `apt-get install default-jdk`
```

> When you use the `apt` package manager, software is downloaded from a central area.

Reviewer: ```suggestion
When you use the `apt`,`yum`, `pacman` or any other package manager, software is downloaded from a central repository. Apart from central repositories there are private ones, and they can be added to the list of repositories for the package manager to look for software.
```

The distribution of focus throughout these modules will be `Ubuntu`. It is designed to be simplistic enough for beginners, but not limited so hinder the potential of a more advanced user.

## Tutorial

There is no tutorial for this module.

## Exercises

> Now conduct some of your own research about different distributions of Linux.

Reviewer: ```suggestion
Now conduct some of your own research about what are the more popular distributions of Linux.
```

<details>

> <summary>Examples</summary>

Reviewer: could you add:

- mint
- elementary os
- pop os

as additional popular OS pls
